### DISCUSSION TITLE
Reactive transactions with mutiny

### DISCUSSION THREAD

**Discussion by exmoria** [jOOQ/jOOQ#14627] [Q&A] [answered]
Just wondering if there are any examples demonstrating the usage of mutiny with jooq? I'm having some troubles trying to satisfy the compiler when working with the mutiny (https://github.com/smallrye/smallrye-mutiny) APIs.

No issues when using project reactor:
```java
create.transactionPublisher(tx -> Mono
      .from(tx.dsl()
        .insertInto(USER_RECORD, USER_RECORD.USERNAME, USER_RECORD.PASSWORD)
        .values("admin", "admin")));
```

And the equivalent, invalid code snippet using the mutiny API:
```java
create.transactionPublisher(tx -> Uni
      .createFrom().publisher(tx.dsl()
        .insertInto(USER_RECORD, USER_RECORD.USERNAME, USER_RECORD.PASSWORD)
        .values("admin", "admin")));
```

```
no instance(s) of type variable(s) T exist so that io.smallrye.mutiny.Uni<T> conforms to org.reactivestreams.Publisher<T>)
```
<br>

I will add that plain old select and insert operations work just fine with mutiny:
```java
var query = create
        .select(USER_RECORD)
        .from(USER_RECORD)
        .where(USER_RECORD.USERNAME.eq(authInfo.getUsername()));

Uni.createFrom()
  .publisher(query)
  .map(r -> r.get(USER_RECORD.USERNAME))
  .subscribe() ...
```

**Comment by lukaseder** [comment, marked as answer]
jOOQ's `DSLContext::transactionPublisher` expects you to pass it a `Publisher`. I don't know why `Uni` doesn't extend `Publisher`, given that `Multi` does. ~Looks like an oversight to me~ (It's never an oversight 😅, it's by design, different opinion from the reactor folks). Probably worth requesting here: https://github.com/smallrye/smallrye-mutiny/issues/966

On the other hand, the `Publisher` type that mutiny supports is the JDK one (`Flow.Publisher`). jOOQ, for historic reasons, still works with the `org.reactivestreams.Publisher` type, so you'd still have to bridge those types as well.

I don't think there are any show stoppers, but given that Mutiny is a bit less RS interoperable than reactor, I suspect you'll be writing much more of this glue code on your end.

**Reply by exmoria** [reply]
Aha, apparently ```Uni``` is not a publisher, so one would need to use ```Multi```. I'd also need to use an adapter library like: https://smallrye.io/smallrye-mutiny-zero/1.0.0/flow-adapters/

So the code snippet from my original post becomes:

```java
 create.transactionPublisher(tx -> AdaptersToReactiveStreams
      .publisher(Multi.createFrom().publisher(tx.dsl()
        .insertInto(USER_RECORD, USER_RECORD.USERNAME, USER_RECORD.PASSWORD)
        .values("admin", "admin"))));
```

Which works but, ugh, this is getting a bit messy, so I think I'll just stick to using reactor. Anyhow, thanks for taking a look at my question. I really appreciate it.

**Reply by lukaseder** [reply]
Yeah, using reactor is more straightforward. You'll have it on your classpath anyway, given that most R2DBC drivers use reactor as a dependency...
